[mlir][Tosa] fix fp16/bf16 support for Clamp min/max attributes

fabrizio-indirli opened the pull request:
In TOSA MLIR dialect, fix the definition of the Clamp op to
accept fp16 & bf16 datatype for the min_fp and max_fp attributes.
Add ClampOp verifier to check attributes types compatibility.
Add related test cases in Tosa/ops.mlir.

fabrizio-indirli committed:
[mlir][Tosa] fix fp16/bf16 support for Clamp min/max attributes

In TOSA MLIR dialect, fix the definition of the Clamp op to
accept fp16 & bf16 datatype for the min_fp and max_fp attributes.
Add ClampOp verifier to check attributes types compatibility.
Add related test cases in Tosa/ops.mlir.

Signed-off-by: Fabrizio Indirli <Fabrizio.Indirli@arm.com>

fabrizio-indirli commented:
@eric-k256 @GeorgeARM
Hey, would be glad to get your feedback on this PR when you have time :)
(not sure if I should add someone else as well ...)

GeorgeARM commented:
Looks good. My only thought is if this is too restrictive. I know that tosa defines only `bf16`, `fp16`, `fp32` at the moment but I think that is probably a good idea to allow any float type and just check that the input type, the output type and the attributes type if float match. @eric-k256 what is your take on this?

eric-k256 commented:
A verifier would be nice to have to check attribute/in/out alignment. I believe that the issue with adding all float types is that bfloat16 doesn't show as a normal float type. Maybe all float + bf16 would work.

fabrizio-indirli commented:
Thank you both for the reviews!
My understanding is that the storage class `mlir::FloatAttr ` accepts also Bf16, but in MLIR the restricted attribute `FloatAttrBase`  pre-defined in _CommonAttrConstraints.td_ (and used, for example, to define F32Attr etc.) does not support Bf16.
Thus we need to define our own attribute type. 
We can create a generic Float attribute if we want, though for now Tosa accepts only Fp16/Bf16/Fp32. We could do this:
- In the TOSA dialect (e.g. calling it Tosa_FloatAttr), as done here
- In _mlir/IR/CommonAttrConstrainst.td_, so that potentially other dialects could use this. However this may require other approvals as we would be modifying the generic IR? Moreover something similar already seems to exist with Builtin_FloatAttr, although I was not able to use it here.

Any thoughts on which solution would be better?

fabrizio-indirli commented:
@GeorgeARM @eric-k256 
I updated the patch so that _min_fp_ & _max_fp_ are a generic Float attribute (defined as `Tosa_FloatAttr` in _TosaTypesBase.td_ ) and added a verifier to check that the types are aligned.
In particular, it checks that:
- Input and output datatype are the same
- If datatype is float, checks that the min_fp type == max_fp type, and that this attributes' type is the same of the input/output datatype OR has an higher bitwidth.
  Supporting this last case (min/max_fp with an higher precision than the input/output float type) should allow to retain the interface unchanged and to [pass existing tests that had "f32" attributes and "f16"/"bf16" input/output datatype](https://github.com/llvm/llvm-project/blob/21e1b13f3384b875bd2205a736570320cb020f3e/mlir/test/Dialect/Tosa/canonicalize.mlir#L50C5-L50C5)

GeorgeARM commented on mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td:
```diff
@@ -380,15 +380,17 @@ def Tosa_ClampOp : Tosa_ElementwiseOp<"clamp"> {
   );
 
   let hasCanonicalizer = 1;
+
```
[nitpick]: no space needed

fabrizio-indirli replied:
fixed

GeorgeARM commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -266,6 +266,32 @@ LogicalResult tosa::AvgPool2dOp::verify() {
+      llvm::cast<ShapedType>(getOutput().getType()).getElementType();
+  unsigned dataTypeBitWidth = inputDT.getIntOrFloatBitWidth();
+
+  if (!(inputDT.getTypeID() == outputDT.getTypeID()))
```
Can't we just compare the IO element types?

fabrizio-indirli replied:
Fixed

fabrizio-indirli commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -266,6 +266,32 @@ LogicalResult tosa::AvgPool2dOp::verify() {
 }
 
+LogicalResult tosa::ClampOp::verify() {
+  mlir::Type inputDT =
```
fixed

GeorgeARM reviewed (commented):
LGTM
